### PR TITLE
Refactor QgsConnectionPool[Group] and it's descendants

### PR DESCRIPTION
**Pull request by reflectored**
## Description
This refactoring job is related to https://github.com/qgis/QGIS-Enhancement-Proposals/issues/272 and the original pull request for the AWS Redshift driver https://github.com/qgis/QGIS/pull/53018.

  - Move qgsConnectionPool_[connectionCreate|connectionDestroy|invalidateConnection|connectionIsValid|connectionToName]   into their appropriate abstract classes and declare them as pure virual.
  - Implement the destructor and the functions above for all providers.
  - Refactor QgsPostgresConnPool[Group] to QgsSQLConnPool[Group], in order to reuse the template for future SQL based providers (AWS Redshift).
  - Refactor Postgres connection settings to use the latest [settings API](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html#settings).

### CODE REVIEW

**Comment by reflectored** [comment]
@nyalldawson @jef-n I started the refactoring work we have discussed. I noticed some redundant code with the connection pool so I started here.

File: src/core/providers/ogr/qgsogrconnpool.cpp [hunk 1]
```diff
@@ -18,6 +18,43 @@
...
+    papszOpenOptions = CSLAddString( papszOpenOptions,
+                                     option.toUtf8().constData() );
+  }
+  c->ds = QgsOgrProviderUtils::GDALOpenWrapper( fullPath.toUtf8().constData(), false, papszOpenOptions, nullptr );
+  CSLDestroy( papszOpenOptions );
+  c->path = connInfo;
+  c->valid = true;
+}
+
+void QgsOgrConnPoolGroup::connectionDestroy( QgsOgrConn *c )
```
Comment (elpaso) [inline src/core/providers/ogr/qgsogrconnpool.cpp:43]: same (I won't repeat for the other occurrences)
```suggestion
void QgsOgrConnPoolGroup::connectionDestroy( QgsOgrConn *connection )
```

File: src/core/providers/ogr/qgsogrconnpool.cpp [hunk 2]
```diff
@@ -18,6 +18,43 @@
...
+{
+  c = new QgsOgrConn;
+
+  const QVariantMap parts = QgsOgrProviderMetadata().decodeUri( connInfo );
+  const QString fullPath = parts.value( QStringLiteral( "vsiPrefix" ) ).toString()
+                           + parts.value( QStringLiteral( "path" ) ).toString()
+                           + parts.value( QStringLiteral( "vsiSuffix" ) ).toString();
+  const QStringList openOptions = parts.value( QStringLiteral( "openOptions" ) ).toStringList();
+  char **papszOpenOptions = nullptr;
+  for ( const QString &option : openOptions )
```
Comment (elpaso) [inline src/core/providers/ogr/qgsogrconnpool.cpp:32]:
```suggestion
  for ( const QString &option : std::as_const( openOptions ) )
```

I know it's redundant, but it makes it explicit and it's a noop anyway.

File: src/core/providers/ogr/qgsogrconnpool.h [hunk 3]
```diff
@@ -92,13 +50,31 @@ class QgsOgrConnPoolGroup : public QObject, public QgsConnectionPoolGroup<QgsOgr
...
+    ~QgsOgrConnPoolGroup() override
+    {
+      for ( const Item &item : std::as_const( conns ) )
+      {
+        destroyOgrConn( item.c );
+      }
+    }
+
+    void connectionCreate( const QString &connInfo, QgsOgrConn *&c ) override;
+    void connectionDestroy( QgsOgrConn *c ) override;
```
Comment (elpaso) [inline src/core/providers/ogr/qgsogrconnpool.h:62]:
```suggestion
    void connectionDestroy( QgsOgrConn *connection ) override;
```
... and other methods too

File: src/core/qgsconnectionpool.h [hunk 4]
```diff
@@ -76,13 +68,28 @@ class QgsConnectionPoolGroup
...
-        qgsConnectionPool_ConnectionDestroy( item.c );
-      }
-    }
+    virtual ~QgsConnectionPoolGroup() = 0;
+
+    /**
+     * Classes that implement this template must implement their own connection functions.
+     * Create a connection \a c based on connection information \a connInfo.
+     */
+    virtual void connectionCreate( const QString &connInfo, T &c ) = 0;
```
Comment (elpaso) [inline src/core/qgsconnectionpool.h:77]: ... and other occurrences too
```suggestion
    virtual void connectionCreate( const QString &connInfo, T &connection ) = 0;
```

**Review by elpaso — COMMENTED** [review]
Thank you for working on this refactoring, it was definitely time to do it.

**Comment by reflectored** [comment]
@elpaso thanks for the review, I have addressed all of your comments!

File: src/core/providers/qgssqlconnpool.h [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+#include "qgslogger.h"
+
+/**
+ * \ingroup core
+ * \brief Template class for SQL based provider connection pool group.
+ * \note not available in Python bindings.
+ * \since QGIS 3.32
+ */
+template<typename T>
+class QgsSQLConnPoolGroup : public QgsConnectionPoolGroup<T *>
```
Comment (elpaso) [inline src/core/providers/qgssqlconnpool.h:30]:
```suggestion
class QgsSqlConnPoolGroup : public QgsConnectionPoolGroup<T *>
```

We follow QT conventions about acronyms case: https://docs.qgis.org/3.28/en/docs/developers_guide/codingstandards.html#functions which applies to function names and classes as well.

File: src/core/qgsconnectionpool.h [hunk 6]
```diff
@@ -212,43 +223,46 @@ class QgsConnectionPoolGroup
...
     }
 
   protected:
 
-    QString connInfo;
-    QStack<Item> conns;
-    QList<T> acquiredConns;
-    QMutex connMutex;
-    QSemaphore sem;
+    QString connectionInfo;
```
Comment (elpaso) [inline src/core/qgsconnectionpool.h:254]:
```suggestion
    QString mConnectionInfo;
```

all class members should be prefixed with `m`, this also applies to all members below.

File: src/providers/postgres/qgsconnectionconfigurator.h [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+    {
+      QgsSettings settings;
+      return settings.value( "/" + T::mConnectionTypeName + "/connections/" + connName + "/projectsInDatabase", false ).toBool();
+    }
+
+    static void deleteConnection( const QString &connName )
+    {
+      QgsSettings settings;
+
+      QString key = "/" + T::mConnectionTypeName + "/connections/" + connName;
```
Comment (elpaso) [inline src/providers/postgres/qgsconnectionconfigurator.h:153]: I know it's boring but all string literals should be wrapped by QStringLiteral or QLatin1String.

File: src/core/providers/qgssqlconnpool.h (hunk 5 above)
Comment (reflectored) [reply src/core/providers/qgssqlconnpool.h:30]: Should I change it to QgsSql**Connection**PoolGroup to comply completely with the convention?

Comment (elpaso) [reply src/core/providers/qgssqlconnpool.h:30]: Yes please, the more readable the better. And thank you for your patience.

**Comment by reflectored** [comment]
@elpaso not sure if the spell_check error is related. Would you mind taking another look?

File: src/providers/hana/qgshanaconnectionpool.h [hunk 8]
```diff
@@ -23,44 +23,48 @@
...
+      : QgsConnectionPoolGroup<QgsHanaConnection*>( name )
+    {
+      initTimer( this );
+    }
+
+    ~QgsHanaConnectionPoolGroup() override
+    {
+      for ( const Item &item : std::as_const( mConnections ) )
+      {
+        delete item.connection;
```
Comment (elpaso) [inline src/providers/hana/qgshanaconnectionpool.h:42]: Just wondering if for this kind of loops a simple call to qDeleteAll would be a better choice https://doc.qt.io/qt-6/qtalgorithms.html#qDeleteAll-1

File: src/providers/postgres/qgsconnectionconfigurator.h [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+class QgsConnectionConfigurator
+{
+    static const QString connections_prefix;
+    static const QString connection_name_prefix;
+
+  public:
+    static QStringList connectionList()
+    {
+      QgsSettings settings;
+      settings.beginGroup( connections_prefix );
```
Comment (elpaso) [inline src/providers/postgres/qgsconnectionconfigurator.h:38]: The settings hierarchy has been recently refactored, and it is exposed to users in the advanced configuration options, you might want to check with @3nids for the best approach here.

File: src/providers/postgres/qgsconnectionconfigurator.h [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+
+    static QgsDataSourceUri connUri( const QString &connName )
+    {
+      QgsDebugMsgLevel( "theConnName = " + connName, 2 );
+
+      QgsSettings settings;
+
+      QString key = connections_prefix + QStringLiteral( "/%1" ).arg( connName );
+
+      QString service = settings.value( key + QLatin1String( "/service" ) ).toString();
```
Comment (elpaso) [inline src/providers/postgres/qgsconnectionconfigurator.h:62]:
```suggestion
      const QString service = settings.value( key + QLatin1String( "/service" ) ).toString();
```

I think these can all be const (except for port).

As a general rule, if a variable can be const, mark it as const.

File: src/providers/postgres/qgsconnectionconfigurator.h [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+      {
+        username = settings.value( key + QLatin1String( "/username" ) ).toString();
+
+        if ( settings.value( key + QLatin1String( "/save" ) ).toString() == QLatin1String( "true" ) )
+        {
+          password = settings.value( key + QLatin1String( "/password" ) ).toString();
+        }
+      }
+
+      QString authcfg = settings.value( key + QLatin1String( "/authcfg" ) ).toString();
```
Comment (elpaso) [inline src/providers/postgres/qgsconnectionconfigurator.h:97]:
```suggestion
      const QString authcfg = settings.value( key + QLatin1String( "/authcfg" ) ).toString();
```

File: src/providers/postgres/qgsconnectionconfigurator.h [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+      settings.remove( key + QLatin1String( "/publicOnly" ) );
+      settings.remove( key + QLatin1String( "/geometryColumnsOnly" ) );
+      settings.remove( key + QLatin1String( "/allowGeometrylessTables" ) );
+      settings.remove( key + QLatin1String( "/estimatedMetadata" ) );
+      settings.remove( key + QLatin1String( "/saveUsername" ) );
+      settings.remove( key + QLatin1String( "/savePassword" ) );
+      settings.remove( key + QLatin1String( "/save" ) );
+      settings.remove( key + QLatin1String( "/authcfg" ) );
+      settings.remove( key + QLatin1String( "/keys" ) );
+      settings.remove( key );
```
Comment (elpaso) [inline src/providers/postgres/qgsconnectionconfigurator.h:174]: I 'm not sure: can't you just remove key to remove all its children?

**Comment by elpaso** [comment]
> @elpaso not sure if the spell_check error is related. Would you mind taking another look?

Looks good to me, I'll try to restart the failing jobs, sometimes it helps (even if it shouldn't).

**Comment by elpaso** [comment]
> > @elpaso not sure if the spell_check error is related. Would you mind taking another look?
> 
> Looks good to me, I'll try to restart the failing jobs, sometimes it helps (even if it shouldn't).

it didn't help, can you try to rebase against current master and force-push?

File: src/providers/postgres/qgsconnectionconfigurator.h (hunk 9 above)
Comment (3nids) [reply src/providers/postgres/qgsconnectionconfigurator.h:38]: yes, you have the documentation here: https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html#settings

you can take example on the OWS connections (WMS, WFS, …) which is mostly identical: https://github.com/qgis/QGIS/blob/b4613b98397a08a054ee712292c15b84e08621b1/src/core/qgsowsconnection.h#L96

File: src/providers/postgres/qgsconnectionconfigurator.h (hunk 12 above)
Comment (3nids) [reply src/providers/postgres/qgsconnectionconfigurator.h:174]: this is properly handled when using the new API

File: src/providers/hana/qgshanaconnectionpool.h (hunk 8 above)
Comment (reflectored) [reply src/providers/hana/qgshanaconnectionpool.h:42]: Notice we delete the element connection inside Item, running qDeleteAll on the whole struct Item will not compile.

Comment (elpaso) [reply src/providers/hana/qgshanaconnectionpool.h:42]: of course, sorry for the noise.

**Comment by reflectored** [comment]
Ok so refactoring PostGIS settings to the new API was not straight forward, when implementing for the future extension to support the Redshift driver. There are 2 translation units (libpostgresprovider and qgis_core/gui), hence it's not possible to template the settings class - otherwise the settings will be defined twice at runtime, during QGIS application start and when libpostgresprovider is loaded. I resolved to constructing a static object of the settings class.

**Comment by 3nids** [comment]
Thanks a lot for updating the settings.

I fail to understand the need of having the settings defined as members of a non static class.
What is the reason of not reusing the same approach than `QgsXyzConnectionSettings` ?

**Comment by reflectored** [comment]
> than

I have started the refactoring job of the PostGIS driver to enable extending it to support the AWS Redshift driver, which uses the same libpq interface.

QgsPostgresConnectionsSettings are identical to (future) Redshift settings, the only difference is the driver name and default port number. The [original Redshift driver PR](https://github.com/qgis/QGIS/pull/53018) contains mostly copy pasted code from the Postgres driver. I do not want to copy paste the settings class from Postgres, templating it does not work due to the reason [mentioned above](https://github.com/qgis/QGIS/pull/53307#issuecomment-1640566012), inheriting the class requires exposing the member variables, which beats the point of inheriting to save lines of code.

I thought about it, and actually copying the PostgresConnectionSettings is the way to go, because the Redshift driver settings might diverge in the future. What do you think?

**Comment by 3nids** [comment]
My preferred approach would indeed be the duplication of the settings. Indeed chances of divergence are there and it's better isolated and also avoid the need for the dynamic naming.

**Comment by reflectored** [comment]
@3nids something seems off with the CI, when I run locally the tests e.g. ALL_BUT_PROVIDERS, I do not hit these failures anymore on my latest patch (due to QgsSettingsException). Only some flaky tests fail for other reasons:
```
98% tests passed, 18 tests failed out of 855

Label Time Summary:
POSTGRES    =   8.31 sec*proc (3 tests)

Total Test time (real) = 2632.87 sec

The following tests FAILED:
	  2 - checkGitStatus (Failed)
	 35 - test_core_authmanager (Failed)
	135 - test_core_networkaccessmanager (Subprocess aborted)
	222 - test_core_openclutils (Subprocess aborted)
	278 - test_gui_filedownloader (Subprocess aborted)
	296 - test_gui_queryresultwidget (Subprocess aborted)
	300 - test_3d_3drendering (Failed)
	324 - test_analysis_processing (Failed)
	328 - test_provider_wcsprovider (Failed)
	482 - PyQgsExternalStorageWebDav (Failed)
	483 - PyQgsExternalStorageAwsS3 (Failed)
	600 - PyQgsMapLayerComboBox (Failed)
	603 - PyQgsMapLayerProxyModel (Failed)
	664 - PyQgsProcessExecutablePt1 (Failed)
	665 - PyQgsProcessExecutablePt2 (Failed)
	769 - PyQgsStyleModel (Failed)
	852 - qgis_sipify (Failed)
	854 - qgis_sip_uptodate (Failed)
Errors while running CTest
Output from these tests are in: /home/qgis/build/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
```
Also I struggle to get a container for the MinGW working, I tried manually running on fedora:rawhide, but then the NodeJS version there is too new for the current configurations. Do you have a particular container for MinGW at hand?

**Review by 3nids — COMMENTED** [review]
on the settings front, everything looks good!
thanks a lot for the changes.

**Comment by reflectored** [comment]
@elpaso it passes all tests now! I appreciate your review.

**Comment by reflectored** [comment]
ping @elpaso

**Comment by elpaso** [comment]
> ping @elpaso

Sorry for the late reply, can you please resolve the merge conflicts?

**Pull request closed without merging**